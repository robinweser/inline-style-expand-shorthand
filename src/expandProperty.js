const WHITESPACE_NO_PARENS = /\s+(?=[^)]*?(?:\(|$))/g
const LENGTH_UNIT = /(em|ex|ch|rem|vw|vh|vmin|vmax|cm|mm|q|in|pt|pc|px|dpi|dpcm|dppx|%|auto)$/i
const EDGE_CASES = /^(calc\(|var--)/i
const CUSTOM_PROPERTIES = /var\((--[\w-]+)\)/g

const BORDER_STYLE = /^(dashed|dotted|double|groove|hidden|inset|none|outset|ridge|solid)$/i
const BORDER_WIDTH = /^(thick|medium|think)$/i

// Replace `var(--foo)` with `var--foo` to remove the nested parens issues
function removeCustomProperties(value) {
  if (value.indexOf('var(') === -1) return value
  return value.replace(CUSTOM_PROPERTIES, (match, $1) => `var${$1}`)
}

// Replace `var--foo` with `var(--foo)` to restore original custom properties
function restoreCustomProperties(value) {
  if (value.indexOf('--') === -1) return value
  return value.replace(/var(--[\w-]+)/g, (match, $1) => `var(${$1})`)
}

function parseBorder(value, resolve) {
  const valueWithoutVar = removeCustomProperties(value)
  const values = valueWithoutVar.split(WHITESPACE_NO_PARENS)
  const longhands = {}

  values.forEach(val => {
    if (val.match(BORDER_STYLE) !== null) {
      longhands[resolve('Style')] = restoreCustomProperties(val)
    } else if (
      val.match(BORDER_WIDTH) !== null ||
      val.match(LENGTH_UNIT) !== null ||
      val.match(EDGE_CASES) !== null ||
      val === '0'
    ) {
      longhands[resolve('Width')] = restoreCustomProperties(val)
    } else {
      longhands[resolve('Color')] = restoreCustomProperties(val)
    }
  })

  return longhands
}

function parseCircular(value, resolve) {
  const valueWithoutVar = removeCustomProperties(value)
  const [Top, Right = Top, Bottom = Top, Left = Right] = valueWithoutVar.split(
    WHITESPACE_NO_PARENS
  )

  return {
    [resolve('Top')]: restoreCustomProperties(Top),
    [resolve('Right')]: restoreCustomProperties(Right),
    [resolve('Bottom')]: restoreCustomProperties(Bottom),
    [resolve('Left')]: restoreCustomProperties(Left),
  }
}

var circularExpand = {
  borderWidth: key => 'border' + key + 'Width',
  borderColor: key => 'border' + key + 'Color',
  borderStyle: key => 'border' + key + 'Style',
  padding: key => 'padding' + key,
  margin: key => 'margin' + key,
}

var borderExpand = {
  borderLeft: key => 'borderLeft' + key,
  borderTop: key => 'borderTop' + key,
  borderRight: key => 'borderRight' + key,
  borderBottom: key => 'borderBottom' + key,
  outline: key => 'outline' + key,
}

function parseFlex(value) {
  const valueWithoutVar = removeCustomProperties(value)
  const values = valueWithoutVar.split(WHITESPACE_NO_PARENS)
  const longhands = {}

  values.forEach(val => {
    if (val.match(LENGTH_UNIT) !== null || val.match(EDGE_CASES) !== null) {
      longhands.flexBasis = restoreCustomProperties(val)
    } else {
      if (longhands.flexGrow) {
        longhands.flexShrink = restoreCustomProperties(val)
      } else {
        longhands.flexGrow = restoreCustomProperties(val)
      }
    }
  })

  return longhands
}

export default function expandProperty(property, value) {
  // special expansion for the border property as its 2 levels deep
  if (property === 'border') {
    const longhands = parseBorder(value.toString(), key => 'border' + key)

    var result = {}
    for (let property in longhands) {
      Object.assign(result, expandProperty(property, longhands[property]))
    }

    return result
  }

  if (property === 'flex') {
    return parseFlex(value.toString())
  }

  if (circularExpand[property]) {
    return parseCircular(value.toString(), circularExpand[property])
  }

  if (borderExpand[property]) {
    return parseBorder(value.toString(), borderExpand[property])
  }
}
