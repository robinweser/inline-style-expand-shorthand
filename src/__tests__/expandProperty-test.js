import expandProperty from '../expandProperty'

describe('Expanding property values', () => {
  it('should expand padding/margin correctly', () => {
    expect(expandProperty('padding', '15px 20px 5px 10px')).toMatchSnapshot()
    expect(expandProperty('padding', '15px 20px 5px')).toMatchSnapshot()
    expect(expandProperty('padding', '15px 20px')).toMatchSnapshot()
    expect(expandProperty('padding', '15px')).toMatchSnapshot()
    expect(expandProperty('padding', 0)).toMatchSnapshot()
    expect(expandProperty('padding', '0')).toMatchSnapshot()
    expect(expandProperty('margin', '15px 20px 5px 10px')).toMatchSnapshot()
    expect(
      expandProperty('padding', 'calc(20px - 15px) 10px calc(100% - 5px)')
    ).toMatchSnapshot()
    expect(
      expandProperty('margin', '15px var(--foo) 5px var(--bar)')
    ).toMatchSnapshot()
    expect(
      expandProperty(
        'padding',
        '15px calc(1em * var(--foo)) 5px calc(10px + var(--bar))'
      )
    ).toMatchSnapshot()
  })

  it('should expand flex correctly', () => {
    expect(expandProperty('flex', '0 1 auto')).toMatchSnapshot()
    expect(expandProperty('flex', '2 1 50%')).toMatchSnapshot()
    expect(expandProperty('flex', '3 1 20px')).toMatchSnapshot()
    expect(expandProperty('flex', '3 20px')).toMatchSnapshot()
    expect(expandProperty('flex', '20px 1')).toMatchSnapshot()
    expect(expandProperty('flex', '20px 1 0')).toMatchSnapshot()
    expect(expandProperty('flex', '1 20px 0')).toMatchSnapshot()
    expect(expandProperty('flex', '1 2 var(--foo)')).toMatchSnapshot()
    expect(
      expandProperty('flex', '1 2 calc(100% - var(--bar))')
    ).toMatchSnapshot()
  })

  it('should expand borders/outline correctly', () => {
    expect(expandProperty('border', '1px solid grey')).toMatchSnapshot()
    expect(
      expandProperty('border', 'calc(10px + 1em) solid grey')
    ).toMatchSnapshot()
    expect(
      expandProperty('border', 'calc(var(--foo) + 1em) solid grey')
    ).toMatchSnapshot()
    expect(expandProperty('border', 'solid 1px grey')).toMatchSnapshot()
    expect(
      expandProperty('border', 'solid calc(10px + 1em) grey')
    ).toMatchSnapshot()
    expect(
      expandProperty('border', 'solid calc(var(--foo) + 1em) grey')
    ).toMatchSnapshot()
    expect(expandProperty('border', 'solid grey 1px')).toMatchSnapshot()
    expect(
      expandProperty('border', 'solid grey calc(10px + 1em)')
    ).toMatchSnapshot()
    expect(
      expandProperty('border', 'solid grey calc(var(--foo) + 1em)')
    ).toMatchSnapshot()

    expect(expandProperty('borderLeft', '1px solid grey')).toMatchSnapshot()

    expect(expandProperty('borderTop', '1px solid grey')).toMatchSnapshot()

    expect(expandProperty('borderWidth', '1px 5px 3px 2px')).toMatchSnapshot()
    expect(expandProperty('borderWidth', '1px 5px 3px')).toMatchSnapshot()

    expect(expandProperty('outline', '1px solid grey')).toMatchSnapshot()
    expect(expandProperty('outline', 'var(--bar) solid grey')).toMatchSnapshot()
    expect(
      expandProperty('outline', 'calc(100px + 1em) solid grey')
    ).toMatchSnapshot()
    expect(
      expandProperty('outline', 'calc(100px + var(--foo)) solid grey')
    ).toMatchSnapshot()

    expect(expandProperty('borderLeft', 0)).toMatchSnapshot()
    expect(expandProperty('borderLeft', 'var(--foo)')).toMatchSnapshot()
    expect(
      expandProperty('borderLeft', 'calc(-1 * var(--bar))')
    ).toMatchSnapshot()
  })
})
