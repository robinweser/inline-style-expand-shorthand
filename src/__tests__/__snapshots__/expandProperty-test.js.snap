// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expanding property values should expand borders/outline correctly 1`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 2`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(10px + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(10px + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(10px + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(10px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 3`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(var(--foo) + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(var(--foo) + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(var(--foo) + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(var(--foo) + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 4`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 5`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(10px + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(10px + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(10px + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(10px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 6`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(var(--foo) + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(var(--foo) + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(var(--foo) + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(var(--foo) + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 7`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 8`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(10px + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(10px + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(10px + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(10px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 9`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(var(--foo) + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(var(--foo) + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(var(--foo) + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(var(--foo) + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 10`] = `
Object {
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 11`] = `
Object {
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 12`] = `
Object {
  "borderBottomWidth": "3px",
  "borderLeftWidth": "2px",
  "borderRightWidth": "5px",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 13`] = `
Object {
  "borderBottomWidth": "3px",
  "borderLeftWidth": "5px",
  "borderRightWidth": "5px",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 14`] = `
Object {
  "outlineColor": "grey",
  "outlineStyle": "solid",
  "outlineWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 15`] = `
Object {
  "outlineColor": "grey",
  "outlineStyle": "solid",
  "outlineWidth": "var(--bar)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 16`] = `
Object {
  "outlineColor": "grey",
  "outlineStyle": "solid",
  "outlineWidth": "calc(100px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 17`] = `
Object {
  "outlineColor": "grey",
  "outlineStyle": "solid",
  "outlineWidth": "calc(100px + var(--foo))",
}
`;

exports[`Expanding property values should expand borders/outline correctly 18`] = `
Object {
  "borderLeftWidth": "0",
}
`;

exports[`Expanding property values should expand borders/outline correctly 19`] = `
Object {
  "borderLeftWidth": "var(--foo)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 20`] = `
Object {
  "borderLeftWidth": "calc(-1 * var(--bar))",
}
`;

exports[`Expanding property values should expand flex correctly 1`] = `
Object {
  "flexBasis": "auto",
  "flexGrow": "0",
  "flexShrink": "1",
}
`;

exports[`Expanding property values should expand flex correctly 2`] = `
Object {
  "flexBasis": "50%",
  "flexGrow": "2",
  "flexShrink": "1",
}
`;

exports[`Expanding property values should expand flex correctly 3`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "3",
  "flexShrink": "1",
}
`;

exports[`Expanding property values should expand flex correctly 4`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "3",
}
`;

exports[`Expanding property values should expand flex correctly 5`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "1",
}
`;

exports[`Expanding property values should expand flex correctly 6`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "1",
  "flexShrink": "0",
}
`;

exports[`Expanding property values should expand flex correctly 7`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "1",
  "flexShrink": "0",
}
`;

exports[`Expanding property values should expand flex correctly 8`] = `
Object {
  "flexBasis": "var(--foo)",
  "flexGrow": "1",
  "flexShrink": "2",
}
`;

exports[`Expanding property values should expand flex correctly 9`] = `
Object {
  "flexBasis": "calc(100% - var(--bar))",
  "flexGrow": "1",
  "flexShrink": "2",
}
`;

exports[`Expanding property values should expand padding/margin correctly 1`] = `
Object {
  "paddingBottom": "5px",
  "paddingLeft": "10px",
  "paddingRight": "20px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 2`] = `
Object {
  "paddingBottom": "5px",
  "paddingLeft": "20px",
  "paddingRight": "20px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 3`] = `
Object {
  "paddingBottom": "15px",
  "paddingLeft": "20px",
  "paddingRight": "20px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 4`] = `
Object {
  "paddingBottom": "15px",
  "paddingLeft": "15px",
  "paddingRight": "15px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 5`] = `
Object {
  "paddingBottom": "0",
  "paddingLeft": "0",
  "paddingRight": "0",
  "paddingTop": "0",
}
`;

exports[`Expanding property values should expand padding/margin correctly 6`] = `
Object {
  "paddingBottom": "0",
  "paddingLeft": "0",
  "paddingRight": "0",
  "paddingTop": "0",
}
`;

exports[`Expanding property values should expand padding/margin correctly 7`] = `
Object {
  "marginBottom": "5px",
  "marginLeft": "10px",
  "marginRight": "20px",
  "marginTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 8`] = `
Object {
  "paddingBottom": "calc(100% - 5px)",
  "paddingLeft": "10px",
  "paddingRight": "10px",
  "paddingTop": "calc(20px - 15px)",
}
`;

exports[`Expanding property values should expand padding/margin correctly 9`] = `
Object {
  "marginBottom": "5px",
  "marginLeft": "var(--bar)",
  "marginRight": "var(--foo)",
  "marginTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 10`] = `
Object {
  "paddingBottom": "5px",
  "paddingLeft": "calc(10px + var(--bar))",
  "paddingRight": "calc(1em * var(--foo))",
  "paddingTop": "15px",
}
`;
